### ISSUE TITLE
JwtParser needs a setVerifyingKey API

### ISSUE THREAD

**Issue by rhonnava** [jwtk/jjwt#208]
The token signature validation requires the caller to specify a signing key. It does not make it obvious that this is a verify operation. 

```
JwtParser setSigningKey(byte[] key);

JwtParser setSigningKey(String base64EncodedKeyBytes);

JwtParser setSigningKey(Key key);
```

In RsaSignatureValidator, the isValid method does a signature verification if the passed key is an instance of type PublicKey. If passed an instance of type PrivateKey, it signs the data again and compares signature. 

But at JWTParser layer, the following APIs will provide a more clear description of the operation: 

```
JwtParser setVerifyingKey(byte[] verifyingKey);

JwtParser setVerifyingKey(String base64EncodedVerifyKeyBytes);

JwtParser setVerifyingKey(PublicKey publicKey);
```

However setSigningKey keys has to be retained for backward compatibility.

setVerifyingKey can also help scenarios where a token issuer service is different from a token validator, and the private key need not to be shared among them.

**Comment by djKianoosh** [comment]
I'm not a maintainer/contributor but I like making the distinction between signing and verifying. In a large organization where I'm using this, only one service creates JWT's with the private key, and all other clients only have the public key to verify with. I think this would help with other people's understanding. 

Just my $0.02.

**Comment by lhazlewood** [comment]
This is better self-documenting - I think it's a good addition.  Thanks for the issue!

**Comment by rhonnava** [comment]
@lhazlewood, sorry I did not get that. Do you mean to document this use case or do you mean to add the APIs in the JWTParser layer?

**Comment by tolask** [comment]
Beside the documentation, IMHO, the most important here is that I don't want to share the private key to everybody... Once the token is generated it can even be thrown out. I suppose that only the public key is expected to be shared and used to verify the signature of the token (as I used to do with Jose4j for example).

**Comment by rhonnava** [comment]
@tolask I agree.

**Comment by djKianoosh** [comment]
well, not "thrown out" ;) but yeah locked and sealed for only the server (jwt issuer) to know about. 👍

**Comment by lhazlewood** [comment]
@rhonnava sorry, didn't see your reply.  I meant to say that the method on the parser should probably be renamed to `setVerificationKey` (or similar) to be more accurate.  We can add the method before 1.0, and deprecate the current one.

For the 1.0 release, we can remove the deprecated `setSigningKey` method since we don't need to support backwards compatibility with 0.x releases at that time.

**Comment by lhazlewood** [comment]
Just noting that this isn't as simple as expected: the `SigningKeyResolver` interface, its adapter, and anything else referencing such needs to be refactored.  This is a change that's probably better left for 1.0 (or at least a release after the more urgent 0.10.0 release).

**Comment by lhazlewood** [comment]
This work is complete with `master`'s latest revisions.  The `JwtParserBuilder.verifyWith` and `JwtParserBuilder.keyLocator` methods are clearer in intent.  This will be released in `0.12.0`
